drv/sprot-api/src/lib.rs

```diff
@@ -3,64 +3,266 @@
-use idol_runtime::{Leased, R};
+use idol_runtime::{Leased, LenLimit, RequestError, R};
 use serde::{Deserialize, Serialize};
-use userlib::{sys_send, FromPrimitive};
-use zerocopy::{AsBytes, FromBytes};
-#[cfg(feature = "sink_test")]
-use zerocopy::{ByteOrder, LittleEndian};
-
-/// The canonical SpRot protocol error returned by this API
-//
-// TODO: Audit that each MsgError is used and has some reasonable action.
-// While a diverse set of error codes may be useful for debugging it
-// clutters code that just has to deal with the error.
-// then consider adding a function that translates an error code
-// into the desired action, e.g. InvalidCrc and FlowError should both
-// result in a retry on the SP side and an ErrorRsp on the RoT side.
-// Note: Variant order must be maintained to stay compatible due to SP/RoT
-// version skew during update.
+use sprockets_common::msgs::{
+    RotError as SprocketsError, RotRequestV1 as SprocketsReq,
+    RotResponseV1 as SprocketsRsp,
+};
+use static_assertions::const_assert;
+use userlib::sys_send;
+
+const CRC16: Crc<u16> = Crc::<u16>::new(&CRC_16_XMODEM);
+pub const CRC_SIZE: usize = <u16 as SerializedSize>::MAX_SIZE;
+// XXX ROT FIFO size should be discovered.
+pub const ROT_FIFO_SIZE: usize = 8;
+pub const MAX_BLOB_SIZE: usize = 512;
+pub const MAX_REQUEST_SIZE: usize =
+    Header::MAX_SIZE + ReqBody::MAX_SIZE + MAX_BLOB_SIZE;
+pub const MAX_RESPONSE_SIZE: usize =
+    Header::MAX_SIZE + RspBody::MAX_SIZE + MAX_BLOB_SIZE;
+
+// For simplicity we want to be able to retrieve the header
+// in a maximum of 1 FIFO size read.
+const_assert!(Header::MAX_SIZE <= ROT_FIFO_SIZE);
+
+pub type Request = Msg<ReqBody>;
+pub type Response = Msg<Result<RspBody, SprotError>>;
+
+/// A message header for a request or response
+///
+/// It's important that this header be kept fixed size by limiting the use
+/// of rust types. This allows us to assume that Header::MAX_SIZE is also the
+/// exact size of the header.
+#[derive(Serialize, Deserialize, SerializedSize)]
+pub struct Header {
+    pub protocol: Protocol,
+    pub body_size: u16,
+}
+
+impl Header {
+    fn new(body_size: u16) -> Header {
+        Header {
+            protocol: Protocol::V2,
+            body_size,
+        }
+    }
+}
+
+/// Information about optional blobs appended to the body of messages
+/// before the CRC.
+pub struct BlobInfo {
+    pub offset: usize,
+    pub size: usize,
+}
+
+/// A sprot Msg that flows between the SP and RoT
+///
+/// The message is parameterized by a `ReqBody` or `RspBody`.
+///
+/// Note that `MSG`s do not implement `Serialize`, `Deserialize`, or
+/// `SerializedSize`, as they need to calculate and place a CRC in the buffer.
+/// `Msg`s sometimes include a an offset into the buffer where a binary blob
+/// resides.
+pub struct Msg<T> {
+    pub header: Header,
+    pub body: T,
+
+    // The index into the serialized buffer where an optional binary blob lives
+    pub blob: Option<BlobInfo>,
```

Comment:
tl;dr: Should this be a `&[u8]`?

`BlobInfo` is an offset and size, which other code typically uses to slice some bytes. As an alternative, you might also consider making this field a `&'a [u8]` referencing the original buffer. This would let callers interact with it without additional slicing code.

You could also eliminate the `Option` and use an empty slice to represent "no blob" -- currently I think the structure is redundant, in that I could have `None` or `Some(BlobInfo { size: 0, .. })`

Reply:
I had this as an `&[u8]` originally but ran into lifetime errors on the SP side, since I was trying to hold onto the slice as returned from `hubpack::deserialize` and then mutably write into it on the next spi read. However, that was just an optimization. I can just recreate the slice after all the mutation is done, and I should do that.

Reply:
I changed it to a slice, but I"m getting problems as the slice is part of `self.tx_buf` which is borrowed mutably in the retry loop of `do_send_recv_retries`. I'm trying to work this out, because a slice is a much nicer thing to return here.

Reply:
One option is to put all the spi code: `do_send_recv`, etc... inside its own type or free functions. We had the former in prior code, but I didn't particularly like it.

Reply:
Fixed in [9f8bed9](https://github.com/oxidecomputer/hubris/pull/1279/commits/9f8bed9ad79b519fb73906e97bfa378665990bec)

Reply:
This was not actually fixed. I introduced a bug with an implicit copy. It appears that there may be a compiler bug that is causing the compiler to think that there is mutable and immutable references to the same slice when inside a loop. This appears to happen when the compiler can't tell that the loop terminates. A minimal reproducible example is here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=59c15df3d0a060a01086450d6bf1a780

I'm still working on solving it. Thanks to @cbiffle for the initial example that I was able to modify to reproduce the issue. @sunshowers has asked some rust compiler devs about it. We'll see what they come up with :)